metrics/tempdata/tempdata.go

```diff
@@ -0,0 +1,147 @@
+
+func (t *TempData) StartTime() time.Time {
+	return time.UnixMicro(int64(atomic.LoadUint64(&t.startTime)))
+}
+
+func (t *TempData) RecordIndexerCandidates(eventTime time.Time, candidatesFound uint32) (first bool) {
+	if atomic.CompareAndSwapUint32(&t.indexerCandidates, 0, candidatesFound) {
+		atomic.StoreUint64(&t.indexerCandidateFirstResultTime, uint64(eventTime.UnixMicro()))
+		return true
+	}
+	atomic.AddUint32(&t.indexerCandidates, candidatesFound)
+	return false
+}
+
+func (t *TempData) RecordIndexerFilteredCandidates(candidatesFoundFiltered uint32) (first bool) {
+	if atomic.CompareAndSwapUint32(&t.indexerFiltered, 0, candidatesFoundFiltered) {
+		return true
+	}
+	atomic.AddUint32(&t.indexerFiltered, candidatesFoundFiltered)
+	return false
+}
+
+func (t *TempData) RecordBitswapAttempt() (first bool) {
+	return atomic.CompareAndSwapUint32(&t.bitswapAttempt, 0, 1)
+}
+func (t *TempData) RecordGraphsyncAttempt() (first bool) {
+	return atomic.CompareAndSwapUint32(&t.graphsyncAttempt, 0, 1)
+}
+
+func (t *TempData) RecordTimeToFirstByte(ttfbTime time.Time) (first bool) {
+	return atomic.CompareAndSwapUint64(&t.ttfbTime, 0, uint64(ttfbTime.UnixMicro()))
+}
+
+func (t *TempData) RecordFailure() (first bool) {
+	if atomic.CompareAndSwapUint32(&t.failedCount, 0, 1) {
+		return true
+	}
+
+	atomic.AddUint32(&t.failedCount, 1)
+	return false
+}
+
+func (t *TempData) RecordFinality() {
+	t.timer.Stop()
+}
+
+type FinalValues struct {
+	StartTime                       time.Time
+	IndexerCandidates               uint64
+	IndexerCandidateFirstResultTime time.Time
+	IndexerFiltered                 uint64
+	HasBitswapAttempt               bool
+	HasGraphSyncAttempt             bool
+	TimeToFirstByte                 time.Time
+	FailedCount                     uint64
+}
+
+func (t *TempData) finalValues() FinalValues {
+	return FinalValues{
+		StartTime:                       time.UnixMicro(int64(atomic.LoadUint64(&t.startTime))),
+		IndexerCandidates:               uint64(atomic.LoadUint32(&t.indexerCandidates)),
+		IndexerCandidateFirstResultTime: time.UnixMicro(int64(atomic.LoadUint64(&t.indexerCandidateFirstResultTime))),
+		IndexerFiltered:                 uint64(atomic.LoadUint32(&t.indexerFiltered)),
+		HasBitswapAttempt:               atomic.LoadUint32(&t.bitswapAttempt) > 0,
+		HasGraphSyncAttempt:             atomic.LoadUint32(&t.graphsyncAttempt) > 0,
+		TimeToFirstByte:                 time.UnixMicro(int64(atomic.LoadUint64(&t.ttfbTime))),
+		FailedCount:                     uint64(atomic.LoadUint32(&t.failedCount)),
+	}
+}
+
+func (t *TempData) zeroOut() {
+	atomic.StoreUint64(&t.startTime, 0)
+	atomic.StoreUint32(&t.indexerCandidates, 0)
+	atomic.StoreUint64(&t.indexerCandidateFirstResultTime, 0)
+	atomic.StoreUint32(&t.indexerFiltered, 0)
+	atomic.StoreUint32(&t.bitswapAttempt, 0)
+	atomic.StoreUint32(&t.graphsyncAttempt, 0)
+	atomic.StoreUint64(&t.ttfbTime, 0)
+	atomic.StoreUint32(&t.failedCount, 0)
+}
+
+var tempDataPool = sync.Pool{
```

Comment:
I gather that we're just relying on the sheer volume of traffic to wipe out the errors introduced by GC wiping away our data occasionally, giving us not-quite-accurate finality reports? With a very busy service, I wonder whether GC is going to get really disruptive here and make this very inaccurate?